### PR TITLE
[WIP] Navigation baseline with OVMM habitat task spec

### PR DESCRIPTION
**Pull request by yvsriram**
## Motivation and Context
 1. Using OVMM task spec on habitat-lab to load episodes
 2. Work habitat-lab v0.2.3 (hydra) configs 
 
TODOs
- [] Separate Env for obs_preprocess functions etc (different from ObjectNav's Env)
- [] Finalize planner outputs (waypoint) and actions (teleport)
- [] Add habitat-lab changes to object_rearrange branch
- [] Update instructions

## Types of changes
`python projects/habitat_objectnav/eval_episode.py`

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: projects/habitat_objectnav/config_utils.py [hunk 1]
```diff
@@ -3,25 +3,30 @@
...
+from omegaconf import DictConfig
+from habitat_baselines.config.default import get_config as get_habitat_config
+import os
 
+def get_config(path: str, opts: Optional[list] = None):
+    config = get_habitat_config(path)
+    return config, ""
 
-def get_config(path: str, opts: Optional[list] = None) -> Tuple[Config, str]:
+def get_config_old(path: str, opts: Optional[list] = None):
```
Comment (cpaxton) [inline projects/habitat_objectnav/config_utils.py:14]: @theophilegervet @exhaustin we have a problem where a lot of the new configs are in a different format. How should we handle this?

Comment (fantaosha) [reply projects/habitat_objectnav/config_utils.py:14]: ok sounds good!

Comment (cpaxton) [reply projects/habitat_objectnav/config_utils.py:14]: previous comment was chris not taosha

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_state.py [hunk 2]
```diff
@@ -139,13 +140,17 @@ def get_visited_map(self, e) -> np.ndarray:
         """Get local visited map for an environment."""
         return np.copy(self.local_map[e, 3, :, :].cpu().float().numpy())
 
+    def get_been_close_map(self, e) -> np.ndarray:
+        """Get map showing regions the agent has been close to"""
+        return np.copy(self.local_map[e, 4, :, :].cpu().float().numpy())
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_state.py:145]: another random hard coded number

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_state.py [hunk 3]
```diff
@@ -139,13 +140,17 @@ def get_visited_map(self, e) -> np.ndarray:
...
+
     def get_semantic_map(self, e) -> np.ndarray:
         """Get local map of semantic categories for an environment."""
         semantic_map = np.copy(self.local_map[e].cpu().float().numpy())
         semantic_map[
-            3 + self.num_sem_categories, :, :
+            4 + self.num_sem_categories, :, :
```
Comment (cpaxton) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_state.py:151]: hard coded numbers

File: src/home_robot_hw/home_robot_hw/env/visualizer.py [hunk 4]
```diff
@@ -139,11 +140,11 @@ def visualize(
             )
         self.last_xy = (curr_x, curr_y)
 
-        semantic_map += 6
+        semantic_map += 7
```
Comment (cpaxton) [inline src/home_robot_hw/home_robot_hw/env/visualizer.py:143]: hard coded values

File: src/home_robot_hw/home_robot_hw/env/visualizer.py [hunk 5]
```diff
@@ -139,11 +140,11 @@ def visualize(
...
 
         # Obstacles, explored, and visited areas
         no_category_mask = (
-            semantic_map == 6 + self.num_sem_categories - 1
+            semantic_map == 7 + self.num_sem_categories - 1
```
Comment (cpaxton) [inline src/home_robot_hw/home_robot_hw/env/visualizer.py:147]: hard coded values

**Review by cpaxton — COMMENTED** [review]
Main issue: lots of hard coded numbers that will make this hard to debug later on. I would really like to put these as global or class variables, or in a constants.py file or something.

File: projects/habitat_objectnav/config_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yvsriram) [reply projects/habitat_objectnav/config_utils.py]: Removed

File: projects/habitat_objectnav/configs/agent/floorplanner_eval.yaml [hunk 6]
```diff
@@ -10,20 +8,20 @@ GROUND_TRUTH_SEMANTICS: 1 # 1: use ground-truth semantics (for debugging / ablat
...
 AGENT:
   max_steps: 500          # maximum number of steps before stopping an episode
-  panorama_start: 1       # 1: turn around 360 degrees when starting an episode, 0: don't
+  panorama_start: 0       # 1: turn around 360 degrees when starting an episode, 0: don't
 
   SEMANTIC_MAP:
-    semantic_categories: mukul_indoor # map semantic channel categories ("coco_indoor", "longtail_indoor", "mukul_indoor")
-    num_sem_categories: 35            # number of map semantic channel categories (16, 257, 35)
+    semantic_categories: rearrange # map semantic channel categories ("coco_indoor", "longtail_indoor", "mukul_indoor")
+    num_sem_categories: 5    # number of map semantic channel categories (16, 257, 35)
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply projects/habitat_objectnav/configs/agent/floorplanner_eval.yaml:24]: Yes, [misc, obj, start_recep, goal_recep, misc]

File: projects/habitat_objectnav/eval_episode.py [hunk 7]
```diff
@@ -13,34 +13,42 @@
...
+from omegaconf import DictConfig, OmegaConf
 
 from home_robot.agent.objectnav_agent.objectnav_agent import ObjectNavAgent
 from home_robot_sim.env.habitat_objectnav_env.habitat_objectnav_env import (
     HabitatObjectNavEnv,
 )
 
 if __name__ == "__main__":
-    config_path = "configs/agent/floorplanner_eval.yaml"
+    config_path = "rearrange/modular_nav.yaml"
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply projects/habitat_objectnav/eval_episode.py:24]: This is relative to `habitat-lab/habitat-baselines/habitat-baselines/config` folder

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py [hunk 8]
```diff
@@ -309,7 +309,7 @@ def _update_local_map_and_pose(
 
         agent_view = torch.zeros(
             batch_size,
-            obs_channels,
+            5 + self.num_sem_categories,
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:331]: This was a bug. This happened to work because `obs_channels` was same as the number of channels in local map earlier.
`obs` has the egocentric observations, while `agent_view` here is supposed to contain the top-down local maps.

**Comment by yvsriram** [comment]
> Main issue: lots of hard coded numbers that will make this hard to debug later on. I would really like to put these as global or class variables, or in a constants.py file or something.

Yes, I agree. I too felt it hard to debug. Also, this will reduce the redundancy of changes required. I am converting all of these to constants.

**Comment by cpaxton** [comment]
Add a comment maybe?

File: projects/habitat_objectnav/eval_episode.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/habitat_objectnav/eval_episode.py:24]: ok got it! so we don't need to do anything here? i'll resolve

File: projects/habitat_objectnav/configs/agent/floorplanner_eval.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpaxton) [reply projects/habitat_objectnav/configs/agent/floorplanner_eval.yaml:24]: can you add a comment on this

**Review by cpaxton — APPROVED** [review]
lgtm once tests are passing

File: src/home_robot_hw/home_robot_hw/env/visualizer.py [hunk 9]
```diff
@@ -176,6 +177,17 @@ def visualize(
         semantic_map_vis = semantic_map_vis.convert("RGB")
         semantic_map_vis = np.flipud(semantic_map_vis)
         semantic_map_vis = semantic_map_vis[:, :, [2, 1, 0]]
+
+        # overlay the regions the agent has been close to
+        been_close_map = np.flipud(np.rint(been_close_map) == 1)
```
[the remark this answers is not in the document]
Comment (yvsriram) [reply src/home_robot_hw/home_robot_hw/env/visualizer.py]: Added

File: src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py [hunk 10]
```diff
@@ -75,7 +75,12 @@ def __init__(self, config, device_id: int = 0):
 
     @torch.no_grad()
     def prepare_planner_inputs(
-        self, obs: torch.Tensor, pose_delta: torch.Tensor, goal_category: torch.Tensor
+        self,
+        obs: torch.Tensor,
+        pose_delta: torch.Tensor,
+        object_goal_category=None,
+        recep_goal_category=None,
+        camera_pose=None,
```
Comment (exhaustin) [inline src/home_robot/home_robot/agent/objectnav_agent/objectnav_agent.py:83]: Type hints

File: src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py
Comment (exhaustin) [inline src/home_robot/home_robot/mapping/semantic/categorical_2d_semantic_map_module.py:1]: Use `MC.NON_SEM_CHANNELS` instead of using the magic number `5 + self.num_sem_categories` in multiple places.
Can also simplify by `self.num_channels = MC.NON_SEM_CHANNELS + self.num_sem_categories`

File: src/home_robot/home_robot/navigation_planner/discrete_planner.py [hunk 11]
```diff
@@ -167,26 +167,24 @@ def plan(
...
 
+        stg_x, stg_y = short_term_goal
+        angle_st_goal = math.degrees(math.atan2(stg_x - start[0], stg_y - start[1]))
+        angle_agent = start_o % 360.0
+        if angle_agent > 180:
+            angle_agent -= 360
+
+        relative_angle = (angle_agent - angle_st_goal) % 360.0
+        if relative_angle > 180:
+            relative_angle -= 360
```
Comment (exhaustin) [inline src/home_robot/home_robot/navigation_planner/discrete_planner.py:178]: Make this angle normalization logic into a helper function, ideally within `home_robot.utils`.

**Review by exhaustin — APPROVED** [review]
Looks good, some minor nits.
